File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.cpp
#include <string>
#include <vector>

using namespace std;

int N;
int count = 0;
// std::vector<int> v;

void DFS(int n, int now, int target, vector<int> &numbers) {
    if (n == N) {
        if (now == target) {
            count++;
        }
        return;
    }
    for (int i = 0; i < 2; i++) {
        // v.emplace_back(i);
        if (i == 0) {
            DFS(n+1, now-numbers[n], target, numbers);
        } else {
            DFS(n+1, now+numbers[n], target, numbers);
        }
        // v.pop_back();
    }
}

int solution(vector<int> numbers, int target) {
    N = numbers.size();
    DFS(0, 0, target, numbers);
    return count;
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
import java.util.*;

class Solution {
    static int how = 0;
    
    public void dfs(int count, int now, int[] numbers, int target) {
        if (count == numbers.length) {
            if (now == target) {
                how++;
            }
            return;
        }
        dfs(count+1, now+numbers[count], numbers, target);
        dfs(count+1, now-numbers[count], numbers, target);    
    }
    
    public int solution(int[] numbers, int target) {
        int answer = 0;
        dfs(0, 0, numbers, target);
        answer = how;
        return answer;
    }
}